From 9a8db163e491ab2c480619692fd9bd5426d3031e Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Thu, 16 Oct 2014 22:17:49 +0800
Subject: [PATCH 1435/1594] MLK-9618-3 usb: chipidea: host: add port reset for
 USB OTG HNP

commit 5a63aca9c698a2eca09473c675ad2c40a62a27d1 from
git://git.freescale.com/imx/linux-2.6-imx.git

This patch adds host port reset implementation for HNP timing requirement,
which will be called when B host starts enumuration.

Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 5df2a65c073dcdf227f1eb340a686a589db5009b)
---
 drivers/usb/chipidea/host.c |   33 +++++++++++++++++++++++++++++++++
 1 files changed, 33 insertions(+), 0 deletions(-)

diff --git a/drivers/usb/chipidea/host.c b/drivers/usb/chipidea/host.c
index 67f1db6..2485482c 100644
--- a/drivers/usb/chipidea/host.c
+++ b/drivers/usb/chipidea/host.c
@@ -121,6 +121,38 @@ static int ci_imx_ehci_bus_resume(struct usb_hcd *hcd)
 	return 0;
 }
 
+#ifdef CONFIG_USB_OTG
+
+static int ci_start_port_reset(struct usb_hcd *hcd, unsigned port)
+{
+	struct ehci_hcd *ehci = hcd_to_ehci(hcd);
+	u32 __iomem *reg;
+	u32 status;
+
+	if (!port)
+		return -EINVAL;
+	port--;
+	/* start port reset before HNP protocol time out */
+	reg = &ehci->regs->port_status[port];
+	status = ehci_readl(ehci, reg);
+	if (!(status & PORT_CONNECT))
+		return -ENODEV;
+
+	/* khubd will finish the reset later */
+	if (ehci_is_TDI(ehci))
+		ehci_writel(ehci, status | (PORT_RESET & ~PORT_RWC_BITS), reg);
+	else
+		ehci_writel(ehci, status | PORT_RESET, reg);
+
+	return 0;
+}
+
+#else
+
+#define ci_start_port_reset    NULL
+
+#endif
+
 /* The below code is based on tegra ehci driver */
 static int ci_imx_ehci_hub_control(
 	struct usb_hcd	*hcd,
@@ -519,6 +551,7 @@ int ci_hdrc_host_init(struct ci_hdrc *ci)
 		ci_ehci_hc_driver.bus_resume = ci_imx_ehci_bus_resume;
 		ci_ehci_hc_driver.hub_control = ci_imx_ehci_hub_control;
 	}
+	ci_ehci_hc_driver.start_port_reset = ci_start_port_reset;
 
 	return 0;
 }
-- 
1.7.5.4

